ally/scraper.py

```diff
@@ -0,0 +1,171 @@
+        acct_containers = self.driver.find_elements_by_class_name('accounts-individual-container')
+
+        for container in acct_containers:
+            spans = container.find_elements_by_tag_name('span')
+            for span in spans:
+                if '...%s' % self.account in span.text:
+                    link = WebDriverWait(container, 10).until(
+                        EC.element_to_be_clickable((By.TAG_NAME, 'a'))
+                    )
+                    link.click()
+                    return
+
+        raise Exception('Account %s not found', self.account)
+
+    def get_transactions(self):
+        transactions = []
+
+        WebDriverWait(self.driver, 10).until(
+            EC.visibility_of_element_located((By.CLASS_NAME, 'transactions-history-table'))
+        )
+        transaction_tables = self.driver.find_elements_by_class_name('transactions-history-table')
+        for table in transaction_tables[1:]:  # skip first table which is empty
+            rows = table.find_elements_by_tag_name('tr')
+
+            for row in rows[1:]:  # skip header row
+                cols = row.find_elements_by_tag_name('td')
+                if len(cols) != 4:
+                    continue
+
+                if cols[1].get_attribute('class') != 'description':
+                    continue
+
+                # All transactions for current month processed, maybe
+                # TODO: make this better
+                if cols[0].text == '':
+                    return transactions
+
+                if cols[0].text != 'Pending':
+                    posted_at = datetime.strptime(cols[0].text, '%b %d, %Y')
+                else:
+                    posted_at = None
+
+                transactions.append({'posted_at': posted_at,
+                                     'description': cols[1].text.strip(),
+                                     'amount': cols[2].text,
+                                     'balance': cols[3].text})
+
+        return transactions
+
+    def run(self):
+        self.login()
+        self.goto_account_page()
+        return self.get_transactions()
+
+
+def save_to_db(cursor, transaction):
+    stmt = """
+        INSERT INTO ally_transactions
+        (posted_at, description, amount, balance) VALUES
+        (%(posted_at)s, %(description)s, %(amount)s, %(balance)s)
+        RETURNING id
+    """
+    cursor.execute(stmt, transaction)
+    return cursor.fetchone()[0]
+
+
+def main(args, password):
+    driver = webdriver.Remote(
+        command_executor=os.environ['SELENIUM_ADDRESS'],
+        desired_capabilities=DesiredCapabilities.CHROME
+    )
+    driver.implicitly_wait(10)  # seconds
+
+    try:
+        ally = AllyScraper(driver, args.username, password, args.account)
+        transactions = ally.run()
+    finally:
+        driver.close()
+
+    if not transactions:
+        sys.exit('No transactions found')
+
+    if args.save:
+        conn = psycopg2.connect(host=os.environ['POSTGRES_HOST'],
```

Comment:
consider moving the database stuff to another module

Reply:
done

Reply:
done